Unexported map value causes go-cmp to panic

Input:

func TestBreakCmp(t *testing.T) {

	type unexported int
	type Foo struct {
		m map[string]unexported
	}

	f1 := &Foo{m: map[string]unexported{"foo": 1, "bar": 2}} // One key will work. Multiple causes panic.
	f2 := &Foo{m: map[string]unexported{"bar": 3, "baz": 4}}

	// Passes!
	cmp.Diff(f1, f2, cmp.Comparer(func(_, _ Foo) bool {
		return true
	}))

	// Panics!
	cmp.Diff(f1, f2, cmp.Comparer(func(_, _ Foo) bool {
		return false
	}))
}

Output:

panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
	panic: ,reflect.Value.Interface: cannot return value obtained from unexported field or method

goroutine 6 [running]:
testing.tRunner.func1(0xc0420a40f0)
	C:/Go/src/testing/testing.go:711 +0x2d9
panic,(0x578e40, 0x5d5660)
	C:/Go/src/runtime/panic.go:491 +0x291
,reflect.valueInterface(0x578e40, 0xc0421065b0, 0xb8, ,0x1, 0x4, 0x2)
	C:/Go/src/reflect/value.go:936, +0x1e4
reflect.Value.Interface(0x578e40, 0xc0421065b0,, 0xb8, 0xc0420fe3e0, 0xc042136000)
	C:/Go/src/reflect/value.go,:925 +0x4b
github.com/google/go-cmp/cmp/internal/value.SortKeys(0xc042102060,, 0x2, 0x2, 0xc042102060, 0x2, ,0x2)
	D:/work/go/src/github.com/google/go-cmp/cmp/internal/value/sort.go:27 +0x16d
,github.com/google/go-cmp/cmp/internal/value.formatAny(0x587e20, 0xc0420026d8, 0x1b5,, 0x1010100, 0x0, 0x25, 0x688ce0)
,	D:/work/go/src/github.com/google/go-cmp/cmp/internal/value/format.go:132 +0x1a38
github.com/google/go-cmp/cmp/internal/value.formatAny,(0x590c40, 0xc0420026c0, 0x199, 0xc001010101, ,0x0, 0xc042100012, 0x9)
	D:/work/go/src/github.com/google/go-cmp/cmp/internal/value/format.go:,153 +0xea0
github.com/google/go-cmp/cmp/internal/value.Format(0x590c40, 0xc0420026c0, 0x199, 0x4bf901, 0xc042100012, 0x9)
	D:/work/go/src/github.com/google/go-cmp/cmp/internal/value/format.go:30 +0x6d
github.com/google/go-cmp/cmp.(*defaultReporter).Report(0xc042038c40, 0x590c40,, 0xc042002680, 0x199, 0x590c40, 0xc0420026c0, 0x199, ,0x590c00, 0xc0420028c0, 0x2, ...)
	D:/work/go/src/github.com/google/go-cmp/cmp/reporter.go:,34 +0xf7
github.com/google/go-cmp/cmp.(*state).report(0xc042082310, 0x583f00,, 0x590c40, 0xc042002680, 0x199, 0x590c40, ,0xc0420026c0, 0x199)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go:492, +0xce
github.com/google/go-cmp/cmp.(*comparer).apply(0xc042070630, 0xc042082310,, 0x590c40, 0xc042002680, 0x199, 0x590c40, 0xc0420026c0,, 0x199, 0xc0420026c0)
	D:/work/go/src/github.com/google/go-cmp/cmp/options.go:315, +0x103
github.com/google/go-cmp/cmp.(*state).tryOptions(0xc042082310, 0x590c40, ,0xc042002680, 0x199, 0x590c40, 0xc0420026c0, 0x199,, 0x688ce0, 0x590c40, 0x199)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go:307 +0x154
github.com/google/go-cmp/cmp.(*state).compareAny,(0xc042082310, 0x590c40, 0xc042002680, 0x199, 0x590c40,, 0xc0420026c0, 0x199)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go:202, +0x343
github.com/google/go-cmp/cmp.(*state).compareAny(0xc042082310, 0x5857e0, ,0xc042002680, 0x16, 0x5857e0, 0xc0420026c0, 0x16)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go:244 +0x1540
github.com/google/go-cmp/cmp.Equal(0x5857e0, 0xc042002680, 0x5857e0, 0xc0420026c0, 0xc042002880, 0x2, 0x2, 0x590c40)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go:86 +0xf8
github.com/google/go-cmp/cmp.Diff(0x5857e0, 0xc042002680, 0x5857e0, 0xc0420026c0, 0xc042002860, 0x2, 0x2, 0xc042093f98, 0x4f6b55)
	D:/work/go/src/github.com/google/go-cmp/cmp/compare.go,:100 +0x13a

--- a/cmp/internal/value/sort.go
+++ b/cmp/internal/value/sort.go
@@ -24,7 +24,7 @@ func SortKeys(vs []reflect.Value) []reflect.Value {
 	// Deduplicate keys (fails for NaNs).
 	vs2 := vs[:1]
 	for _, v := range vs[1:] {
-		if v.Interface() != vs2[len(vs2)-1].Interface() {
+		if isLess(vs2[len(vs2)-1], v) {
 			vs2 = append(vs2, v)
 		}
 	}

--- a/cmp/internal/value/sort_test.go
+++ b/cmp/internal/value/sort_test.go
@@ -132,15 +132,22 @@ func TestSortKeys(t *testing.T) {
 			complex(math.NaN(), math.NaN()): true,
 		},
 		want: []interface{}{
-			math.NaN(), math.NaN(), math.NaN(), math.NaN(),
-			complex(math.NaN(), math.NaN()), complex(math.NaN(), math.NaN()),
-			complex(math.NaN(), 0), complex(math.NaN(), 0), complex(math.NaN(), 0), complex(math.NaN(), 0),
-			complex(0, math.NaN()), complex(0, math.NaN()), complex(0, math.NaN()), complex(0, math.NaN()),
+			math.NaN(),
+			complex(math.NaN(), math.NaN()),
+			complex(math.NaN(), 0),
+			complex(0, math.NaN()),
 		},
 	}}
 
 	for i, tt := range tests {
-		keys := append(reflect.ValueOf(tt.in).MapKeys(), reflect.ValueOf(tt.in).MapKeys()...)
+		// Intentionally pass the map via an unexported field to detect panics.
+		// Unfortunately, we cannot actually test the keys without using unsafe.
+		v := reflect.ValueOf(struct{ x map[interface{}]bool }{tt.in}).Field(0)
+		value.SortKeys(append(v.MapKeys(), v.MapKeys()...))
+
+		// Try again, with keys that have read-write access in reflect.
+		v = reflect.ValueOf(tt.in)
+		keys := append(v.MapKeys(), v.MapKeys()...)
 		var got []interface{}
 		for _, k := range value.SortKeys(keys) {
 			got = append(got, k.Interface())

